File: day3/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func jolts(bank string, nbat int) int {

	s, p := "", 0
	for b := nbat; b > 0; b-- {
		for i := p; i <= len(bank)-b; i++ {
			if bank[i] > bank[p] {
				p = i
			}

		}

		s += string(bank[p])
		p++
	}
	n, _ := strconv.Atoi(s)
	return n

}

func main() {
	file, _ := os.ReadFile("t.txt")

	part1, part2 := 0, 0

	for s := range strings.FieldsSeq(string(file)) {
		part1 += jolts(s, 2)
		part2 += jolts(s, 12)

	}

	fmt.Println(part1)
	fmt.Println(part2)

}

File: day3/main.py
with open("t.txt", "r") as f:
    parts = f.read().strip().split("\n")
ans = 0


def part2(parts):
    ans = 0
    for line in parts:
        nums = [*map(int, line)]
        n = 0

        for i in range(11, -1, -1):
            temp = nums[: len(nums) - i]
            maxi = max(temp)
            idx = temp.index(maxi)
            nums = nums[idx + 1 :]
            n = n * 10 + maxi
        ans += n
    print(ans)


def part1(parts):
    ans = 0

    for bank in parts:
        bank = str(bank)
        n = len(bank)
        best_digit = bank[0]

        dp = [0] * n

        for i in range(1, n):
            pair_val = int(best_digit + bank[i])
            dp[i] = max(dp[i - 1], pair_val)

            if best_digit < bank[i]:
                best_digit = bank[i]

        ans += dp[n - 1]

    print(ans)


part1(parts)
part2(parts)
